### PR TITLE
Initial commit for ES165RD-II support (rebased)

### PR DESCRIPTION
**Pull request by gavanderhoorn**
Rebased version of #12.

Taking over the review from @shaun-edwards.

Added some commits to fixup minor things.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Attribution is retained.

We'll squash-merge this so as to avoid bloating the history.

**Comment by gavanderhoorn** [comment]
Thanks @marip8 for the original PR.

File: motoman_es_support/urdf/es165rdii_macro.xacro [hunk 1]
```diff
@@ -154,56 +162,62 @@
...
+      <limit lower="${radians(-130)}" upper="${radians(80)}" effort="0" velocity="${radians(105)}"/>
     </joint>
-    <joint name="${prefix}joint_u" type="revolute">
+    <joint name="${prefix}joint_3_u" type="revolute">
       <origin xyz="1.150 0 0" rpy="0 0 0"/>
       <parent link="${prefix}link_l"/>
       <child link="${prefix}link_u"/>
       <axis xyz="0 -1 0"/>
-      <limit lower="${-radians(138.0)}" upper="${radians(78.0)}" effort="0" velocity="${radians(105)}"/>
+      <limit lower="${radians(-112)}" upper="${radians(208)}" effort="0" velocity="${radians(105)}"/>
```
Comment (marip8) [inline motoman_es_support/urdf/es165rdii_macro.xacro:188]: Do we want to represent joint limits like this one (whose limits are not defined relative to the previous joint) verbatim per the spec sheet? In the case of this particular joint, I believe its limits change based on the value of joint L and are measured relative to a fixed vertical plane. We cannot represent this with URDF, but the [-138, 78] relative range more accurately (but not totally accurately) represents the physical range of the robot

**Review by marip8 — COMMENTED** [review]
Sidebar: I have tested/used this URDF with a real ES series robot (see open PR #28) so would it make more sense to move this PR (and #28) to the non-experimental repository?

File: motoman_es_support/urdf/es165rdii_macro.xacro (hunk 1 above)
Comment (gavanderhoorn) [reply motoman_es_support/urdf/es165rdii_macro.xacro:188]: I changed this as we use the specced joint limits for all other robot support packages, across all supported vendors.

Doing something different here would mean diverging from that, which would require additional documentation.

We can do this, but I'd like to do it only if necessary.

Perhaps @EricMarcil and/or @ted-miller have an idea about what would be best to do here?

**Comment by gavanderhoorn** [comment]
@marip8 wrote:
> Sidebar: I have tested/used this URDF with a real ES series robot (see open PR #28) so would it make more sense to move this PR (and #28) to the non-experimental repository?

Yes, we could do that.

Would you be willing to resubmit the mentioned PRs to `ros-industrial/motoman`?

I'd just take the "current state" of this PR, as-if squashed. We don't necessarily need all the fixup commits.

**Comment by marip8** [comment]
I will squash these commits and open a new PR with both models on the other repository

**Comment by gavanderhoorn** [comment]
Please do keep the PRs separate. So one PR per variant.

File: motoman_es_support/urdf/es165rdii_macro.xacro (hunk 1 above)
Comment (ted-miller) [reply motoman_es_support/urdf/es165rdii_macro.xacro:188]: This joint's limit does not change based on the value of L. It is measured relative to L and not to a vertical plane.
(There some some 4-axis palletizing robots which are fixed on a vertical plane. But those are mechanically fixed in that orientation.)
Annotation 2020-07-30 145834
Annotation 2020-07-30 150359
Annotation 2020-07-30 151643
Annotation 2020-07-30 151707

Comment (marip8) [reply motoman_es_support/urdf/es165rdii_macro.xacro:188]: I haven't come across those limits in any of the spec sheets that I have for our 200R*-120 model, but it seems to make sense in the context of the work envelope diagram.

The robot in our lab definitely has a higher positive joint limit than the 79 degrees stated in the document above and can actually achieve an "elbow-down" configuration like in the screenshot below. 

image

@ted-miller is this a non-standard feature? (I think you and @EricMarcil might have changed that configuration for us a long time ago) Or is this capability partially defined by the [-112 deg, 208 deg] joint limits on the spec sheet?

Comment (ted-miller) [reply motoman_es_support/urdf/es165rdii_macro.xacro:188]: I believe that your robot is a non-standard configuration. It is possible to modify the pulse limits for an axis. But, I expect that any modifications would impact the allowable moment and inertia for that axis. 
Untitled

Comment (gavanderhoorn) [reply motoman_es_support/urdf/es165rdii_macro.xacro:188]: Thanks for the info @ted-miller.

@marip8: this is actually why we always just use the specs in the datasheet / documentation.

It's perfectly possible for someone to configure different limits, but they'd have to do that in a fork.

Comment (marip8) [reply motoman_es_support/urdf/es165rdii_macro.xacro:188]: Okay, I'm on board with using the datasheet specs

**Pull request closed without merging**